llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -1438,10 +1438,8 @@ std::string NVPTXTargetLowering::getPrototype(
 
     if (!Outs[OIdx].Flags.isByVal()) {
       if (IsTypePassedAsArray(Ty)) {
-        const CallInst *CallI = cast<CallInst>(&CB);
         Align ParamAlign =
-            getAlign(*CallI, i + AttributeList::FirstArgIndex)
-                .value_or(getFunctionParamOptimizedAlign(F, Ty, DL));
+            getArgumentAlignment(&CB, Ty, i + AttributeList::FirstArgIndex, DL);
```

Comment:
One of the things `getFunctionParamOptimizedAlign` is doing is `increase their alignment in a way that
ensures that we can effectively vectorize their loads & stores.` 
https://github.com/llvm/llvm-project/blob/694b132177a96a61dac62b3e3d2989a063feafa7/llvm/lib/Target/NVPTX/NVPTXISelLowering.h#L468

I think that with this change we may lose this increase and that will be a potential performance regression.

While we're on the subject of changes, it would be great to add a test verifying that this increased alignment on PTX-level function arguments remains in place.

Reply:
If there is no alignment metadata or the call is indirect, the `getArgumentAlignment` path ultimately uses `getFunctionParamOptimizedAlign`. If the call is indirect it uses the ABI type alignment, which `getFunctionParamOptimizedAlign` does as well, so we shouldn't lose out on opportunities for increased alignments.

Unfortunately, I don't think we can test that the increased alignment remains right now. We only emit prototypes for indirect calls, which would use the ABI type alignment.

Reply:
I thik we do have a test that demonstrates the increased alignment:

https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/NVPTX/param-vectorize-device.ll#L140

Does it pass with your patch?

Reply:
All tests pass! By being unable to test with the increased alignment, I was referring to case where we have a `.callprototype` with e.g. `.param .align 16` being impossible both before and after this patch due to indirect calls causing `getFunctionParamOptimizedAlign` to default to ABI alignment. So the cases this patch would affect already could not use the increased alignment.

Reply:
Ah, I see. LGTM.